File: wrappers/python/tests/test_mcp.py
"""Tests for the MCP middleware adapter (`checkrd.mcp`).

The official `mcp` SDK is intentionally NOT installed in test deps —
the wrapper is structurally typed against the MCP shape, so the tests
use hand-rolled stand-ins that match the SDK's surface (`call_tool`,
`read_resource`, `get_prompt`, `list_tools`) and verify wrapping
behaviour without pulling in the real package.
"""

from __future__ import annotations

import asyncio
import json
from typing import Any

import pytest

from checkrd.engine import WasmEngine
from checkrd.exceptions import CheckrdPolicyDenied
from checkrd.mcp import wrap_mcp_client, wrap_mcp_server

ALLOW_ALL = json.dumps({"agent": "t", "default": "allow", "rules": []})
DENY_ALL = json.dumps({"agent": "t", "default": "deny", "rules": []})
DENY_SEARCH = json.dumps(
    {
        "agent": "t",
        "default": "allow",
        "rules": [
            {
                "name": "block-search-tool",
                "deny": {
                    "method": ["POST"],
                    "url": "mcp/tools/search",
                },
            },
        ],
    },
)


class StubClient:
    """Minimal MCP-shaped client. Methods are async, mirroring the
    real SDK; we record every call for inspection."""

    def __init__(self) -> None:
        self.call_tool_calls: list[dict[str, Any]] = []
        self.read_resource_calls: list[dict[str, Any]] = []
        self.get_prompt_calls: list[dict[str, Any]] = []
        self.list_tools_calls = 0

    async def call_tool(
        self, name: str, arguments: Any = None,
    ) -> dict[str, Any]:
        self.call_tool_calls.append({"name": name, "arguments": arguments})
        return {"content": [{"type": "text", "text": "ok"}]}

    async def read_resource(self, uri: str) -> dict[str, Any]:
        self.read_resource_calls.append({"uri": uri})
        return {"contents": [{"uri": uri, "text": "hi"}]}

    async def get_prompt(
        self, name: str, arguments: Any = None,
    ) -> dict[str, Any]:
        self.get_prompt_calls.append({"name": name, "arguments": arguments})
        return {"messages": []}

    async def list_tools(self) -> dict[str, Any]:
        self.list_tools_calls += 1
        return {"tools": []}


class TestWrapMcpClientAllowPath:
    def test_call_tool_forwards_under_allow_all(self) -> None:
        engine = WasmEngine(ALLOW_ALL, "test")
        raw = StubClient()
        client = wrap_mcp_client(
            raw, engine=engine, enforce=True, agent_id="test",
        )
        result = asyncio.run(
            client.call_tool("search", arguments={"q": "x"}),
        )
        assert result == {"content": [{"type": "text", "text": "ok"}]}
        assert raw.call_tool_calls == [{"name": "search", "arguments": {"q": "x"}}]

    def test_read_resource_forwards(self) -> None:
        engine = WasmEngine(ALLOW_ALL, "test")
        raw = StubClient()
        client = wrap_mcp_client(
            raw, engine=engine, enforce=True, agent_id="test",
        )
        asyncio.run(client.read_resource(uri="file://x"))
        assert raw.read_resource_calls == [{"uri": "file://x"}]

    def test_get_prompt_forwards(self) -> None:
        engine = WasmEngine(ALLOW_ALL, "test")
        raw = StubClient()
        client = wrap_mcp_client(
            raw, engine=engine, enforce=True, agent_id="test",
        )
        asyncio.run(client.get_prompt(name="greet", arguments={"who": "world"}))
        assert raw.get_prompt_calls == [
            {"name": "greet", "arguments": {"who": "world"}},
        ]

    def test_list_tools_forwards(self) -> None:
        engine = WasmEngine(ALLOW_ALL, "test")
        raw = StubClient()
        client = wrap_mcp_client(
            raw, engine=engine, enforce=True, agent_id="test",
        )
        asyncio.run(client.list_tools())
        assert raw.list_tools_calls == 1


class TestWrapMcpClientDenyPath:
    def test_default_deny_blocks_in_enforce_mode(self) -> None:
        engine = WasmEngine(DENY_ALL, "test")
        raw = StubClient()
        client = wrap_mcp_client(
            raw, engine=engine, enforce=True, agent_id="test",
        )
        with pytest.raises(CheckrdPolicyDenied):
            asyncio.run(client.call_tool("search"))
        assert raw.call_tool_calls == []

    def test_specific_tool_denied_via_url_matcher(self) -> None:
        engine = WasmEngine(DENY_SEARCH, "test")
        raw = StubClient()
        client = wrap_mcp_client(
            raw, engine=engine, enforce=True, agent_id="test",
        )
        with pytest.raises(CheckrdPolicyDenied):
            asyncio.run(client.call_tool("search"))
        assert raw.call_tool_calls == []
        # Other tools still pass:
        asyncio.run(client.call_tool("fetch"))
        assert raw.call_tool_calls == [{"name": "fetch", "arguments": None}]

    def test_observe_only_forwards_despite_deny(self) -> None:
        engine = WasmEngine(DENY_ALL, "test")
        raw = StubClient()
        client = wrap_mcp_client(
            raw, engine=engine, enforce=False, agent_id="test",
        )
        result = asyncio.run(client.call_tool("search"))
        assert result == {"content": [{"type": "text", "text": "ok"}]}
        assert raw.call_tool_calls == [{"name": "search", "arguments": None}]


class TestWrapMcpClientProxyTransparency:
    def test_unknown_attributes_pass_through(self) -> None:
        engine = WasmEngine(ALLOW_ALL, "test")

        class WithExtras(StubClient):
            custom_field = "preserved"

            async def custom_method(self) -> str:
                return "result"

        raw = WithExtras()
        client = wrap_mcp_client(
            raw, engine=engine, enforce=True, agent_id="test",
        )
        assert client.custom_field == "preserved"
        assert asyncio.run(client.custom_method()) == "result"


class TestWrapMcpServerHandlerWrapping:
    def test_decorator_factory_wraps_async_handler(self) -> None:
        engine = WasmEngine(ALLOW_ALL, "test")
        registered: list[Any] = []
        user_handler_calls: list[Any] = []

        class StubServer:
            def call_tool(self) -> Any:  # decorator factory
                def decorator(fn: Any) -> Any:
                    registered.append(fn)
                    return fn
                return decorator

        srv = wrap_mcp_server(
            StubServer(), engine=engine, enforce=True, agent_id="test",
        )

        async def my_tool(request: Any) -> dict[str, Any]:
            user_handler_calls.append(request)
            return {"ok": True}

        # Apply the decorator explicitly so we keep a reference to the
        # original `my_tool` — the `@deco` sugar would reassign the
        # name to the decorator's return value.
        decorated = srv.call_tool()(my_tool)

        # Register fired exactly once; the registered fn is the
        # policy-wrapped one, not the raw handler.
        assert len(registered) == 1
        wrapped = registered[0]
        assert wrapped is not my_tool
        # Since StubServer's decorator returns fn unchanged, `decorated`
        # is `wrapped` — the user's `my_tool` name (if they'd used the
        # sugar) would point at the policy-wrapped fn too.
        assert decorated is wrapped

        # Calling the wrapped handler runs through policy + delegates
        # to the real user handler with the original args.
        class _Req:
            class params:  # noqa: D106
                name = "my_tool"
                arguments = {"q": "x"}

        req = _Req()
        result = asyncio.run(wrapped(req))
        assert result == {"ok": True}
        assert user_handler_calls == [req]

    def test_decorator_factory_blocks_under_deny(self) -> None:
        engine = WasmEngine(DENY_ALL, "test")
        registered: list[Any] = []
        user_handler_ran = False

        class StubServer:
            def call_tool(self) -> Any:
                def decorator(fn: Any) -> Any:
                    registered.append(fn)
                    return fn
                return decorator

        srv = wrap_mcp_server(
            StubServer(), engine=engine, enforce=True, agent_id="test",
        )

        async def blocked_tool(request: Any) -> dict[str, Any]:  # noqa: ARG001
            nonlocal user_handler_ran
            user_handler_ran = True  # pragma: no cover
            return {"ok": True}

        srv.call_tool()(blocked_tool)
        wrapped = registered[0]

        class _Req:
            class params:  # noqa: D106
                name = "blocked_tool"
                arguments = None

        with pytest.raises(CheckrdPolicyDenied):
            asyncio.run(wrapped(_Req()))
        assert not user_handler_ran

    def test_set_request_handler_wraps_handler(self) -> None:
        engine = WasmEngine(ALLOW_ALL, "test")
        registered: list[Any] = []

        class StubServer:
            def set_request_handler(self, schema: Any, handler: Any) -> None:  # noqa: ARG002
                registered.append(handler)

        srv = wrap_mcp_server(
            StubServer(), engine=engine, enforce=True, agent_id="test",
        )

        async def my_handler(request: Any) -> dict[str, Any]:  # noqa: ARG001
            return {"ok": True}

        srv.set_request_handler({"fake": "schema"}, my_handler)
        wrapped = registered[0]
        assert wrapped is not my_handler

        class _Req:
            class params:  # noqa: D106
                name = "anything"
                arguments = None

        result = asyncio.run(wrapped(_Req()))
        assert result == {"ok": True}

File: wrappers/javascript/tests/integrations/test_mcp.ts
/**
 * MCP middleware tests — both the client wrapper (intercepts calls
 * before they reach the server) and the server wrapper (intercepts
 * handler registrations before user code runs).
 *
 * The MCP SDK itself is not installed in devDependencies (it's a
 * runtime-optional peer dep when customers actually use MCP), so
 * these tests use hand-rolled structural stand-ins that match the
 * shapes the real SDK emits.
 */
import { describe, expect, it, vi } from "vitest";

import { WasmEngine } from "../../src/engine.js";
import { CheckrdPolicyDenied } from "../../src/exceptions.js";
import {
  wrapMcpClient,
  wrapMcpServer,
  type McpClientLike,
  type McpRequestHandler,
  type McpServerLike,
} from "../../src/integrations/_mcp.js";

const ALLOW_ALL = JSON.stringify({ agent: "t", default: "allow", rules: [] });
const DENY_ALL = JSON.stringify({ agent: "t", default: "deny", rules: [] });
const DENY_SEARCH = JSON.stringify({
  agent: "t",
  default: "allow",
  rules: [
    {
      name: "block-search-tool",
      deny: {
        method: ["POST"],
        url: "mcp/tools/search",
      },
    },
  ],
});

function stubClient(): McpClientLike & {
  callTool: ReturnType<typeof vi.fn>;
  readResource: ReturnType<typeof vi.fn>;
  getPrompt: ReturnType<typeof vi.fn>;
  listTools: ReturnType<typeof vi.fn>;
} {
  return {
    callTool: vi.fn(async (_params: { name: string; arguments?: unknown }) => ({
      content: [{ type: "text", text: "ok" }],
    })),
    readResource: vi.fn(async (_params: { uri: string }) => ({
      contents: [{ uri: _params.uri, text: "hi" }],
    })),
    getPrompt: vi.fn(async (_params: { name: string }) => ({ messages: [] })),
    listTools: vi.fn(async () => ({ tools: [] })),
  };
}

describe("wrapMcpClient — allow path", () => {
  it("forwards callTool when policy allows", async () => {
    const engine = new WasmEngine(ALLOW_ALL, "test");
    const client = stubClient();
    const wrapped = wrapMcpClient(client, {
      engine,
      enforce: true,
      agentId: "test",
    });
    const result = await wrapped.callTool({ name: "search", arguments: { q: "x" } });
    expect(result).toEqual({ content: [{ type: "text", text: "ok" }] });
    expect(client.callTool).toHaveBeenCalledOnce();
  });

  it("forwards readResource when policy allows", async () => {
    const engine = new WasmEngine(ALLOW_ALL, "test");
    const client = stubClient();
    const wrapped = wrapMcpClient(client, {
      engine,
      enforce: true,
      agentId: "test",
    });
    await wrapped.readResource!({ uri: "file://x" });
    expect(client.readResource).toHaveBeenCalledOnce();
  });

  it("forwards listTools when policy allows", async () => {
    const engine = new WasmEngine(ALLOW_ALL, "test");
    const client = stubClient();
    const wrapped = wrapMcpClient(client, {
      engine,
      enforce: true,
      agentId: "test",
    });
    await wrapped.listTools!();
    expect(client.listTools).toHaveBeenCalledOnce();
  });
});

describe("wrapMcpClient — deny path", () => {
  it("throws CheckrdPolicyDenied on default-deny policy under enforce=true", async () => {
    const engine = new WasmEngine(DENY_ALL, "test");
    const client = stubClient();
    const wrapped = wrapMcpClient(client, {
      engine,
      enforce: true,
      agentId: "test",
    });
    await expect(
      wrapped.callTool({ name: "search" }),
    ).rejects.toBeInstanceOf(CheckrdPolicyDenied);
    expect(client.callTool).not.toHaveBeenCalled();
  });

  it("blocks a specific tool via URL matcher", async () => {
    const engine = new WasmEngine(DENY_SEARCH, "test");
    const client = stubClient();
    const wrapped = wrapMcpClient(client, {
      engine,
      enforce: true,
      agentId: "test",
    });
    // search is denied
    await expect(
      wrapped.callTool({ name: "search" }),
    ).rejects.toBeInstanceOf(CheckrdPolicyDenied);
    expect(client.callTool).not.toHaveBeenCalled();
    // other tools still work
    await wrapped.callTool({ name: "fetch" });
    expect(client.callTool).toHaveBeenCalledOnce();
  });

  it("under enforce=false, forwards the call but fires telemetry", async () => {
    const engine = new WasmEngine(DENY_ALL, "test");
    const client = stubClient();
    const wrapped = wrapMcpClient(client, {
      engine,
      enforce: false,
      agentId: "test",
    });
    const result = await wrapped.callTool({ name: "search" });
    expect(result).toEqual({ content: [{ type: "text", text: "ok" }] });
    expect(client.callTool).toHaveBeenCalledOnce();
  });
});

describe("wrapMcpClient — proxy transparency", () => {
  it("forwards unknown properties unchanged", () => {
    const engine = new WasmEngine(ALLOW_ALL, "test");
    const client = {
      ...stubClient(),
      customField: "preserved",
      customMethod: () => "result",
    };
    const wrapped = wrapMcpClient(client as unknown as McpClientLike, {
      engine,
      enforce: true,
      agentId: "test",
    });
    expect((wrapped as unknown as { customField: string }).customField).toBe("preserved");
    expect((wrapped as unknown as { customMethod: () => string }).customMethod()).toBe("result");
  });
});

describe("wrapMcpServer — handler wrapping", () => {
  it("registerTool handler runs after policy check under allow-all", async () => {
    const engine = new WasmEngine(ALLOW_ALL, "test");
    let registered: McpRequestHandler | null = null;
    const server: McpServerLike = {
      registerTool: vi.fn(
        (_name: string, _config: unknown, handler: McpRequestHandler) => {
          registered = handler;
          return undefined;
        },
      ),
    };
    const userHandler = vi.fn(async () => ({ result: "ok" }));

    const wrapped = wrapMcpServer(server, {
      engine,
      enforce: true,
      agentId: "test",
    });
    wrapped.registerTool!("search", { schema: {} }, userHandler);

    expect(server.registerTool).toHaveBeenCalledOnce();
    // Now invoke the registered handler (which is the policy-wrapped version).
    const out = await registered!({ params: { name: "search", arguments: {} } });
    expect(out).toEqual({ result: "ok" });
    expect(userHandler).toHaveBeenCalledOnce();
  });

  it("registerTool handler blocks user handler under default-deny", async () => {
    const engine = new WasmEngine(DENY_ALL, "test");
    let registered: McpRequestHandler | null = null;
    const server: McpServerLike = {
      registerTool: vi.fn(
        (_name: string, _config: unknown, handler: McpRequestHandler) => {
          registered = handler;
          return undefined;
        },
      ),
    };
    const userHandler = vi.fn(async () => ({ result: "ok" }));

    const wrapped = wrapMcpServer(server, {
      engine,
      enforce: true,
      agentId: "test",
    });
    wrapped.registerTool!("search", { schema: {} }, userHandler);

    await expect(
      registered!({ params: { name: "search" } }),
    ).rejects.toBeInstanceOf(CheckrdPolicyDenied);
    expect(userHandler).not.toHaveBeenCalled();
  });

  it("setRequestHandler wraps the handler the same way", async () => {
    const engine = new WasmEngine(ALLOW_ALL, "test");
    let registered: McpRequestHandler | null = null;
    const server: McpServerLike = {
      setRequestHandler: vi.fn((_schema: unknown, handler: McpRequestHandler) => {
        registered = handler;
        return undefined;
      }),
    };
    const userHandler = vi.fn(async () => ({ ok: true }));

    const wrapped = wrapMcpServer(server, {
      engine,
      enforce: true,
      agentId: "test",
    });
    wrapped.setRequestHandler!({ fake: "schema" }, userHandler);

    await registered!({ params: { name: "anything" } });
    expect(userHandler).toHaveBeenCalledOnce();
  });
});
